### PR TITLE
Save options as json

### PR DESCRIPTION
**Pull request by OzoneH3**
Is there a way to formate the saved json? It's pretty much unreadable right now.

### CODE REVIEW

**Comment by BevapDin** [comment]
> Is there a way to formate the saved json? It's pretty much unreadable right now.

The constructor of `JsonOut` has a second parameter: `pretty_format` or similar, set it to `true` (example can be found in input.cpp).

Instead of `fout << serialize();` (which serializes into a temporary string and writes that out), you can write directly to the file. `JsonOut` takes a generic `ostream` as parameter (and `fout` is such a stream):
```C++
JsonOut jout( fout, true );
serialize( jout );
```

----

Doesn't the redundant redundancy of the function names `save_options` and `load_options` (which are names of functions) annoy or at the very least bother you? They operate on an `options_manager` object (aptly received via a `get_options` function), it seems quite clear that they load/save the options.

----

I think you don't need to use `fopen_exclusive` (which, btw. needs a proper call to `fclose_exclusive` inside the catch handler as well, theoretically). The options file are user/player specific, the exclusive file opening is only required for files that may be shared by different players at the same time (shared map files). It's a bit inconsistent as the file is read non-exclusively.

**Comment by OzoneH3** [comment]
Yeah, that looks way better thanks.
Fixed my copy/pasted unknowingness.

Now that you mention it, no, can't complain about any redundancy ocd syndroms, thanks for asking :P. (Changed it anyway, you're welcome)

**Comment by OzoneH3** [comment]
Just saw that the worldoptions are saved as txt as well, will change those too.

**Comment by OzoneH3** [comment]
Save worldoptions as json, done.

**Comment by Coolthulhu** [comment]
Breaks tests

**Comment by OzoneH3** [comment]
Hmm, compiles and links just fine over here?!

**Comment by Coolthulhu** [comment]
OK, was just my old obj file.

Now (after closing the game, but without re-linking it) I'm getting crashes when running the game. options.json looks fine, will check other files.

**Comment by Coolthulhu** [comment]
The crash happens if you run the game and there are worlds with worldoptions.json generated for them.

When running without save directory or when no saves have worldoptions.json, everything runs fine. But since the missing worldoptions.json are generated on game start (not loading), the next time cataclysm is executed, it crashes.

Might be related to OS' handling of files. Win7 x64, but game build with x32.

**Comment by OzoneH3** [comment]
Sounds like Windows dosen't like something in ```get_all_worlds()``` or ```load_world_options()``` everything you describe is working here on Linux.
Maybe the Json loading part in ```load_world_options()```? Thats the only part thats new.

**Comment by Coolthulhu** [comment]
Could be something like binary mode vs text mode.

I'm pretty sure we had something like this happen before, when jsonizing some other kind of data.

File: src/options.cpp [hunk 1]
```diff
@@ -1489,113 +1495,140 @@ void show_options(bool ingame)
...
+bool options_manager::save(bool ingame)
+{
+    bIngame = ingame;
+    const auto savefile = FILENAMES["options"];
+
+    try {
+        std::ofstream fout;
+        fout.exceptions(std::ios::badbit | std::ios::failbit);
+
+        fout.open(savefile.c_str());
```
Comment (BevapDin) [inline src/options.cpp]: You probably need to open the file in binary mode:
```C++
fout.open( ..., std::ios::binary );
```

Comment (OzoneH3) [inline src/options.cpp]: Thanks thats probably it. Only the json load streams seem to need it though after looking through the other load/save functions.

**Comment by kevingranade** [comment]
Looks good to me, but I can't very the windows fix.

**Comment by OzoneH3** [comment]
Finally managed to compile a windows cross build. Crashes without and works with the fix.

**Comment by karlnp** [comment]
You are all heroes for remembering and dealing with Windows' idiosyncrasies.

**Comment by BevapDin** [comment]
Well, looks fine and works fine. One thing: the old options.txt files should be deleted after conversion to JSON. Currently they linger around without any purpose. Even worse: if the user decides to reset their options via deleting of the options file (the JSON file), the game loads the old options.txt files instead of keeping everything at the default.

**Comment by OzoneH3** [comment]
Done, legacy file gets deleted after conversion.

**Comment by Coolthulhu** [comment]
Minor: Autopickup options get converted, but old file is not deleted after conversion.

**Pull request merged**